# Business Case: Netflix - Data Exploration and Visualisation-               Submitted by Anne-DSML DEC -22 MWF 

1. PROBLEM STATEMENT
Neflix is one of the largest OTT platform and it is operating in many countries. And New contents are addded day by day. And in the Exploratory data analaysis , I have analysed the dataset , and the Data analysis , will help to learn, how netflix is successful in content release and types of contents, mostly released , successful genre content , Successful pair of Actor and director movie released monstly. When Analysing the data , we can identify , get to know the data in past, present , and helps to predict the future. As Netflix is already running its business successfully. We can predict, how the business is successful in past years and help to run the business successful and predictable.


Following data analysis have been made in this notebook
1. How has the number of movies released per year changed over the last 20-30 years?
2. Comparison of tv shows vs. movies.
3. What is the best time(day) to launch a TV show?
4. Analysis of actors/directors of different types of shows/movies.
    a) Most casted Actor in Netflix content
    b) Most director contributed to the netflix content
    c) Top directors in different genre
    d) Top Actors in different genre
    e) Successful Actor-Director-genre combo in Netlfix
5. Does Netflix has more focus on TV Shows than movies in recent years
6. Understanding what content is available in different countries
7. Total Content
8. Genre wise movie and TV shows released by year
9. Top 20 countries contributed to Netflix by type
10. Movies Vs TV shows released in last 5 years- Rating wise
11. Movies Vs TV shows released in last 5 years- Genre wise
12. Most content released in month wise

# Importing Libraries

In [533]:
import pandas as pd
import numpy as np
import seaborn as sns
import plotly.express as px
import plotly.graph_objects as go
from plotly.subplots import make_subplots
import matplotlib.pyplot as plt
import ipywidgets as w
from IPython.display import display

# Importing Original dataset

In [534]:
df=pd.read_csv("https://d2beiqkhq929f0.cloudfront.net/public_assets/assets/000/000/940/original/netflix.csv")
df.shape
df.head(5)

show_id     type                  title         director  \
0      s1    Movie   Dick Johnson Is Dead  Kirsten Johnson   
1      s2  TV Show          Blood & Water              NaN   
2      s3  TV Show              Ganglands  Julien Leclercq   
3      s4  TV Show  Jailbirds New Orleans              NaN   
4      s5  TV Show           Kota Factory              NaN   

                                                cast        country  \
0                                                NaN  United States   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
2  Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...            NaN   
3                                                NaN            NaN   
4  Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...          India   

           date_added  release_year rating   duration  \
0  September 25, 2021          2020  PG-13     90 min   
1  September 24, 2021          2021  TV-MA  2 Seasons   
2  September 24, 2021          2021  TV-MA   1 Season   
3  September 24, 2021          2021  TV-MA   1 Season   
4  September 24, 2021          2021  TV-MA  2 Seasons   

                                           listed_in  \
0                                      Documentaries   
1    International TV Shows, TV Dramas, TV Mysteries   
2  Crime TV Shows, International TV Shows, TV Act...   
3                             Docuseries, Reality TV   
4  International TV Shows, Romantic TV Shows, TV ...   

                                         description  
0  As her father nears the end of his life, filmm...  
1  After crossing paths at a party, a Cape Town t...  
2  To protect his family from a powerful drug lor...  
3  Feuds, flirtations and toilet talk go down amo...  
4  In a city of coaching centers known to train I...

# 2. Getting Dataset information -Observations

# shape of the data

In [535]:
df.shape

(8807, 12)

# data types of all the attributes

In [536]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8807 entries, 0 to 8806
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   show_id       8807 non-null   object
 1   type          8807 non-null   object
 2   title         8807 non-null   object
 3   director      6173 non-null   object
 4   cast          7982 non-null   object
 5   country       7976 non-null   object
 6   date_added    8797 non-null   object
 7   release_year  8807 non-null   int64 
 8   rating        8803 non-null   object
 9   duration      8804 non-null   object
 10  listed_in     8807 non-null   object
 11  description   8807 non-null   object
dtypes: int64(1), object(11)
memory usage: 825.8+ KB


# missing value detection

In [537]:
df.isna().sum()

show_id            0
type               0
title              0
director        2634
cast             825
country          831
date_added        10
release_year       0
rating             4
duration           3
listed_in          0
description        0
dtype: int64

In [538]:
print(round((df.isna().sum()/df.shape[0]),4)*100)

show_id          0.00
type             0.00
title            0.00
director        29.91
cast             9.37
country          9.44
date_added       0.11
release_year     0.00
rating           0.05
duration         0.03
listed_in        0.00
description      0.00
dtype: float64


# statistical summary

In [539]:
df.describe()

release_year
count   8807.000000
mean    2014.180198
std        8.819312
min     1925.000000
25%     2013.000000
50%     2017.000000
75%     2019.000000
max     2021.000000

# 3.  Non-Graphical Analysis: Value counts and unique attributes

In [541]:
print("type")
print(df['type'].unique())
print(df['type'].nunique())

type
['Movie' 'TV Show']
2


In [542]:
print("director")
print(df['director'].unique())
print(df['director'].nunique())

director
['Kirsten Johnson' nan 'Julien Leclercq' ... 'Majid Al Ansari'
 'Peter Hewitt' 'Mozez Singh']
4528


In [543]:
print("title")
print(df['title'].unique())
print(df['title'].nunique())

title
['Dick Johnson Is Dead' 'Blood & Water' 'Ganglands' ... 'Zombieland'
 'Zoom' 'Zubaan']
8807


In [544]:
print("cast")
print(df['cast'].unique())
print(df['cast'].nunique())

cast
[nan
 'Ama Qamata, Khosi Ngema, Gail Mabalane, Thabang Molaba, Dillon Windvogel, Natasha Thahane, Arno Greeff, Xolile Tshabalala, Getmore Sithole, Cindy Mahlangu, Ryle De Morny, Greteli Fincham, Sello Maake Ka-Ncube, Odwa Gwanya, Mekaila Mathys, Sandi Schultz, Duane Williams, Shamilla Miller, Patrick Mofokeng'
 'Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabiha Akkari, Sofia Lesaffre, Salim Kechiouche, Noureddine Farihi, Geert Van Rampelberg, Bakary Diombera'
 ...
 'Jesse Eisenberg, Woody Harrelson, Emma Stone, Abigail Breslin, Amber Heard, Bill Murray, Derek Graf'
 'Tim Allen, Courteney Cox, Chevy Chase, Kate Mara, Ryan Newman, Michael Cassidy, Spencer Breslin, Rip Torn, Kevin Zegers'
 'Vicky Kaushal, Sarah-Jane Dias, Raaghav Chanana, Manish Chaudhary, Meghna Malik, Malkeet Rauni, Anita Shabdish, Chittaranjan Tripathy']
7692


In [545]:
print("rating")
print(df['rating'].unique())
print(df['rating'].nunique())

rating
['PG-13' 'TV-MA' 'PG' 'TV-14' 'TV-PG' 'TV-Y' 'TV-Y7' 'R' 'TV-G' 'G'
 'NC-17' '74 min' '84 min' '66 min' 'NR' nan 'TV-Y7-FV' 'UR']
17


In [546]:
print("duration")
print(df['duration'].unique())
print(df['duration'].nunique())

duration
['90 min' '2 Seasons' '1 Season' '91 min' '125 min' '9 Seasons' '104 min'
 '127 min' '4 Seasons' '67 min' '94 min' '5 Seasons' '161 min' '61 min'
 '166 min' '147 min' '103 min' '97 min' '106 min' '111 min' '3 Seasons'
 '110 min' '105 min' '96 min' '124 min' '116 min' '98 min' '23 min'
 '115 min' '122 min' '99 min' '88 min' '100 min' '6 Seasons' '102 min'
 '93 min' '95 min' '85 min' '83 min' '113 min' '13 min' '182 min' '48 min'
 '145 min' '87 min' '92 min' '80 min' '117 min' '128 min' '119 min'
 '143 min' '114 min' '118 min' '108 min' '63 min' '121 min' '142 min'
 '154 min' '120 min' '82 min' '109 min' '101 min' '86 min' '229 min'
 '76 min' '89 min' '156 min' '112 min' '107 min' '129 min' '135 min'
 '136 min' '165 min' '150 min' '133 min' '70 min' '84 min' '140 min'
 '78 min' '7 Seasons' '64 min' '59 min' '139 min' '69 min' '148 min'
 '189 min' '141 min' '130 min' '138 min' '81 min' '132 min' '10 Seasons'
 '123 min' '65 min' '68 min' '66 min' '62 min' '74 min' '131 min' '39 mi

# By analysing unique values, it is understood that columns director,cast, listed_in and country has multiple entries , hence we need to un-nest the data

# 4. Visual Analysis - Univariate, Bivariate after pre-processing of the data

# pre-processing of the data-Pre-processing involves unnesting of the data in columns like Actor, Director, Country,listed_in

In [548]:
#step1: identify
print(df['director'].unique())
print(df['cast'].unique())
print(df['country'].unique())
print(df['listed_in'].unique())

['Kirsten Johnson' nan 'Julien Leclercq' ... 'Majid Al Ansari'
 'Peter Hewitt' 'Mozez Singh']
[nan
 'Ama Qamata, Khosi Ngema, Gail Mabalane, Thabang Molaba, Dillon Windvogel, Natasha Thahane, Arno Greeff, Xolile Tshabalala, Getmore Sithole, Cindy Mahlangu, Ryle De Morny, Greteli Fincham, Sello Maake Ka-Ncube, Odwa Gwanya, Mekaila Mathys, Sandi Schultz, Duane Williams, Shamilla Miller, Patrick Mofokeng'
 'Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabiha Akkari, Sofia Lesaffre, Salim Kechiouche, Noureddine Farihi, Geert Van Rampelberg, Bakary Diombera'
 ...
 'Jesse Eisenberg, Woody Harrelson, Emma Stone, Abigail Breslin, Amber Heard, Bill Murray, Derek Graf'
 'Tim Allen, Courteney Cox, Chevy Chase, Kate Mara, Ryan Newman, Michael Cassidy, Spencer Breslin, Rip Torn, Kevin Zegers'
 'Vicky Kaushal, Sarah-Jane Dias, Raaghav Chanana, Manish Chaudhary, Meghna Malik, Malkeet Rauni, Anita Shabdish, Chittaranjan Tripathy']
['United States' 'South Africa' nan 'India'
 'United States, Ghana, Burkin

# here columns: director, cast, country, listed_in have mutiple entries hence split these columns

In [549]:
#step 2: split
df['director']=df['director'].str.split(",")
df['cast']=df['cast'].str.split(",")
df['country']=df['country'].str.split(",")
df['listed_in']=df['listed_in'].str.split(",")
df.head(5)

show_id     type                  title           director  \
0      s1    Movie   Dick Johnson Is Dead  [Kirsten Johnson]   
1      s2  TV Show          Blood & Water                NaN   
2      s3  TV Show              Ganglands  [Julien Leclercq]   
3      s4  TV Show  Jailbirds New Orleans                NaN   
4      s5  TV Show           Kota Factory                NaN   

                                                cast          country  \
0                                                NaN  [United States]   
1  [Ama Qamata,  Khosi Ngema,  Gail Mabalane,  Th...   [South Africa]   
2  [Sami Bouajila,  Tracy Gotoas,  Samuel Jouy,  ...              NaN   
3                                                NaN              NaN   
4  [Mayur More,  Jitendra Kumar,  Ranjan Raj,  Al...          [India]   

           date_added  release_year rating   duration  \
0  September 25, 2021          2020  PG-13     90 min   
1  September 24, 2021          2021  TV-MA  2 Seasons   
2  September 24, 2021          2021  TV-MA   1 Season   
3  September 24, 2021          2021  TV-MA   1 Season   
4  September 24, 2021          2021  TV-MA  2 Seasons   

                                           listed_in  \
0                                    [Documentaries]   
1  [International TV Shows,  TV Dramas,  TV Myste...   
2  [Crime TV Shows,  International TV Shows,  TV ...   
3                          [Docuseries,  Reality TV]   
4  [International TV Shows,  Romantic TV Shows,  ...   

                                         description  
0  As her father nears the end of his life, filmm...  
1  After crossing paths at a party, a Cape Town t...  
2  To protect his family from a powerful drug lor...  
3  Feuds, flirtations and toilet talk go down amo...  
4  In a city of coaching centers known to train I...

In [550]:
# step 3 : explode
def explode_row(row):
    row['director'] = pd.Series(row['director']).explode().reset_index(drop=True)
    row['cast'] = pd.Series(row['cast']).explode().reset_index(drop=True)
    row['country'] = pd.Series(row['country']).explode().reset_index(drop=True)
    row['listed_in'] = pd.Series(row['listed_in']).explode().reset_index(drop=True)
    return row

In [551]:
df_expl = df.apply(explode_row, axis=1).explode('director').explode('cast').explode('country').explode('listed_in')
df_expl.head(5)

show_id     type                 title         director          cast  \
0      s1    Movie  Dick Johnson Is Dead  Kirsten Johnson           NaN   
1      s2  TV Show         Blood & Water              NaN    Ama Qamata   
1      s2  TV Show         Blood & Water              NaN    Ama Qamata   
1      s2  TV Show         Blood & Water              NaN    Ama Qamata   
1      s2  TV Show         Blood & Water              NaN   Khosi Ngema   

         country          date_added  release_year rating   duration  \
0  United States  September 25, 2021          2020  PG-13     90 min   
1   South Africa  September 24, 2021          2021  TV-MA  2 Seasons   
1   South Africa  September 24, 2021          2021  TV-MA  2 Seasons   
1   South Africa  September 24, 2021          2021  TV-MA  2 Seasons   
1   South Africa  September 24, 2021          2021  TV-MA  2 Seasons   

                listed_in                                        description  
0           Documentaries  As her father nears the end of his life, filmm...  
1  International TV Shows  After crossing paths at a party, a Cape Town t...  
1               TV Dramas  After crossing paths at a party, a Cape Town t...  
1            TV Mysteries  After crossing paths at a party, a Cape Town t...  
1  International TV Shows  After crossing paths at a party, a Cape Town t...

In [552]:
df_expl.shape

(202065, 12)

# writing the un-nested data set into csv file

In [553]:
df_expl.to_csv("df_expl.csv")

# This dataset contains extra spaces(leading and trailing ) in the column entries, hence these spaces must be removed for proper analysis

In [554]:
df_exploded=pd.read_csv("df_expl.csv",skipinitialspace=True)
df_exploded.shape

(202065, 13)

# Analysing unique entries for each column in the dataset

In [555]:
print("type")
print(df_exploded['type'].unique())
print("director")
print(df_exploded['director'].unique())
print("title")
print(df_exploded['title'].unique())
print("cast")
print(df_exploded['cast'].unique())
print("rating")
print(df_exploded['rating'].unique())
print("duration")
print(df_exploded['duration'].unique())
print("country")
print(df_exploded['country'].unique())

type
['Movie' 'TV Show']
director
['Kirsten Johnson' nan 'Julien Leclercq' ... 'Majid Al Ansari'
 'Peter Hewitt' 'Mozez Singh']
title
['Dick Johnson Is Dead' 'Blood & Water' 'Ganglands' ... 'Zombieland'
 'Zoom' 'Zubaan']
cast
[nan 'Ama Qamata' 'Khosi Ngema' ... 'Malkeet Rauni' 'Anita Shabdish'
 'Chittaranjan Tripathy']
rating
['PG-13' 'TV-MA' 'PG' 'TV-14' 'TV-PG' 'TV-Y' 'TV-Y7' 'R' 'TV-G' 'G'
 'NC-17' '74 min' '84 min' '66 min' 'NR' nan 'TV-Y7-FV' 'UR']
duration
['90 min' '2 Seasons' '1 Season' '91 min' '125 min' '9 Seasons' '104 min'
 '127 min' '4 Seasons' '67 min' '94 min' '5 Seasons' '161 min' '61 min'
 '166 min' '147 min' '103 min' '97 min' '106 min' '111 min' '3 Seasons'
 '110 min' '105 min' '96 min' '124 min' '116 min' '98 min' '23 min'
 '115 min' '122 min' '99 min' '88 min' '100 min' '6 Seasons' '102 min'
 '93 min' '95 min' '85 min' '83 min' '113 min' '13 min' '182 min' '48 min'
 '145 min' '87 min' '92 min' '80 min' '117 min' '128 min' '119 min'
 '143 min' '114 min' '118 min' '1

# Datatype for Showid and Date_added must be changes to integer and datetime64[ns] respectively
because: show_id cannot be an string, id will be unique and best possible way to represnt id is integer
data cannot be a string value, it must be in a date format, and day and month can be extracted from the date type

In [556]:
df_exploded['show_id'] = df_exploded['show_id'].str.replace('\D', '',regex=True)
#converting Show_id object dtype to int dtype
df_exploded['show_id'] = df_exploded['show_id'].fillna(0).astype(int)

In [557]:
#converting date_added object dtype to datetime64[ns] dtype so that we can extract month, day from the date
df_exploded['date_added'] = pd.to_datetime(df_exploded['date_added'])

In [558]:
df_exploded.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 202065 entries, 0 to 202064
Data columns (total 13 columns):
 #   Column        Non-Null Count   Dtype         
---  ------        --------------   -----         
 0   Unnamed: 0    202065 non-null  int64         
 1   show_id       202065 non-null  int32         
 2   type          202065 non-null  object        
 3   title         202065 non-null  object        
 4   director      151422 non-null  object        
 5   cast          199916 non-null  object        
 6   country       190062 non-null  object        
 7   date_added    201907 non-null  datetime64[ns]
 8   release_year  202065 non-null  int64         
 9   rating        201998 non-null  object        
 10  duration      202062 non-null  object        
 11  listed_in     202065 non-null  object        
 12  description   202065 non-null  object        
dtypes: datetime64[ns](1), int32(1), int64(2), object(9)
memory usage: 19.3+ MB


# Add a new column to the DataFrame to store the day of the week

In [559]:
df_exploded['Day'] = df_exploded['date_added'].dt.day_name()

# Add a new column to the DataFrame to store the month of the year

In [560]:
df_exploded['month'] = df_exploded['date_added'].dt.month_name()

In [561]:
df_exploded.head(5)

Unnamed: 0  show_id     type                 title         director  \
0           0        1    Movie  Dick Johnson Is Dead  Kirsten Johnson   
1           1        2  TV Show         Blood & Water              NaN   
2           1        2  TV Show         Blood & Water              NaN   
3           1        2  TV Show         Blood & Water              NaN   
4           1        2  TV Show         Blood & Water              NaN   

          cast        country date_added  release_year rating   duration  \
0          NaN  United States 2021-09-25          2020  PG-13     90 min   
1   Ama Qamata   South Africa 2021-09-24          2021  TV-MA  2 Seasons   
2   Ama Qamata   South Africa 2021-09-24          2021  TV-MA  2 Seasons   
3   Ama Qamata   South Africa 2021-09-24          2021  TV-MA  2 Seasons   
4  Khosi Ngema   South Africa 2021-09-24          2021  TV-MA  2 Seasons   

                listed_in                                        description  \
0           Documentaries  As her father nears the end of his life, filmm...   
1  International TV Shows  After crossing paths at a party, a Cape Town t...   
2               TV Dramas  After crossing paths at a party, a Cape Town t...   
3            TV Mysteries  After crossing paths at a party, a Cape Town t...   
4  International TV Shows  After crossing paths at a party, a Cape Town t...   

        Day      month  
0  Saturday  September  
1    Friday  September  
2    Friday  September  
3    Friday  September  
4    Friday  September

In [562]:
df_exploded=df_exploded.drop('Unnamed: 0', axis=1)

In [563]:
df_exploded.head(5)

show_id     type                 title         director         cast  \
0        1    Movie  Dick Johnson Is Dead  Kirsten Johnson          NaN   
1        2  TV Show         Blood & Water              NaN   Ama Qamata   
2        2  TV Show         Blood & Water              NaN   Ama Qamata   
3        2  TV Show         Blood & Water              NaN   Ama Qamata   
4        2  TV Show         Blood & Water              NaN  Khosi Ngema   

         country date_added  release_year rating   duration  \
0  United States 2021-09-25          2020  PG-13     90 min   
1   South Africa 2021-09-24          2021  TV-MA  2 Seasons   
2   South Africa 2021-09-24          2021  TV-MA  2 Seasons   
3   South Africa 2021-09-24          2021  TV-MA  2 Seasons   
4   South Africa 2021-09-24          2021  TV-MA  2 Seasons   

                listed_in                                        description  \
0           Documentaries  As her father nears the end of his life, filmm...   
1  International TV Shows  After crossing paths at a party, a Cape Town t...   
2               TV Dramas  After crossing paths at a party, a Cape Town t...   
3            TV Mysteries  After crossing paths at a party, a Cape Town t...   
4  International TV Shows  After crossing paths at a party, a Cape Town t...   

        Day      month  
0  Saturday  September  
1    Friday  September  
2    Friday  September  
3    Friday  September  
4    Friday  September

# shape of data

In [564]:
df_exploded.shape

(202065, 14)

# data types of all the attributes

In [565]:
df_exploded.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 202065 entries, 0 to 202064
Data columns (total 14 columns):
 #   Column        Non-Null Count   Dtype         
---  ------        --------------   -----         
 0   show_id       202065 non-null  int32         
 1   type          202065 non-null  object        
 2   title         202065 non-null  object        
 3   director      151422 non-null  object        
 4   cast          199916 non-null  object        
 5   country       190062 non-null  object        
 6   date_added    201907 non-null  datetime64[ns]
 7   release_year  202065 non-null  int64         
 8   rating        201998 non-null  object        
 9   duration      202062 non-null  object        
 10  listed_in     202065 non-null  object        
 11  description   202065 non-null  object        
 12  Day           201907 non-null  object        
 13  month         201907 non-null  object        
dtypes: datetime64[ns](1), int32(1), int64(1), object(11)
memory usage: 2

# Identifying null values present in different columns

In [566]:
print(df_exploded.isna().sum())

show_id             0
type                0
title               0
director        50643
cast             2149
country         12003
date_added        158
release_year        0
rating             67
duration            3
listed_in           0
description         0
Day               158
month             158
dtype: int64


In [567]:
print(round((df_exploded.isna().sum()/df_exploded.shape[0]),4)*100)

show_id          0.00
type             0.00
title            0.00
director        25.06
cast             1.06
country          5.94
date_added       0.08
release_year     0.00
rating           0.03
duration         0.00
listed_in        0.00
description      0.00
Day              0.08
month            0.08
dtype: float64


# Statistical summary

In [568]:
df_exploded.describe()

show_id   release_year
count  202065.000000  202065.000000
mean     4373.623487    2013.448950
std      2592.618026       9.013616
min         1.000000    1925.000000
25%      2122.000000    2012.000000
50%      4330.000000    2016.000000
75%      6679.000000    2019.000000
max      8807.000000    2021.000000

# 4.1, 4.2 and 4.3 are done after handling the missing values in the dataset

# DATA CLEANING-Handling missing values

# Analysis will only be as good as the quality of  data . so it is very important to see this step-through carefully.

# As we see, there are'nt any rows, where more than 25% data is missing, so dropping the row is not feasible in this case
we have 3 options to handle these NaN values:
    Option 1: Ignore them
    Option 2: fill with some anonymous value or fill with "Unknown"
    Option 3: Drop the rows
    Option 4: Impute the NaN values by to creating separate group of objects,with these grouping combinations,replace the NaN with the most feasible value from that group only

Option 1,2,3 are very expensive opreations which leads to data loss and proper insights and recommendations cannot be provided.
Hence opting for 4 by imputing the values will be a feasible soultion.

# Steps followed for imputing the values: Option 4:

Step 1: Choose the right combination of objects and create a group
Step 2: find out the best value to be replaced with the NaN
Step 3: Create a dictionary with a key:Value Pair
Step 4: fill the NaN values with Dictionary key:value pair

# Handling Missing Values for Director Column

In [569]:
#create dataframe by grouping the objects type  and rating and get the value_counts for director for  each combination
df_directors=df_exploded.groupby(['type','rating'],as_index=False)['director'].value_counts()
df_directors.head(10)

type  rating              director  count
0  Movie  66 min            Louis C.K.      1
1  Movie  74 min            Louis C.K.      1
2  Movie  84 min            Louis C.K.      1
3  Movie       G          Robert Vince    108
4  Movie       G            Ken Hughes     90
5  Movie       G            Mel Stuart     90
6  Movie       G         John A. Davis     72
7  Movie       G  Juan Pablo Buscarini     66
8  Movie       G       Mark A.Z. Dippé     56
9  Movie       G          Paul Demeyer     56

In [570]:
#for each group get the get the director name whose made maximum movies in that particular group
df_directors_count=df_directors.loc[df_directors.reset_index().groupby(['type','rating'])['count'].idxmax()]
df_directors_count.head(10)

type  rating             director  count
0     Movie  66 min           Louis C.K.      1
1     Movie  74 min           Louis C.K.      1
2     Movie  84 min           Louis C.K.      1
3     Movie       G         Robert Vince    108
50    Movie   NC-17  Abdellatif Kechiche    117
52    Movie      NR       Lars von Trier    120
135   Movie      PG         Raja Gosnell    308
413   Movie   PG-13        Peter Jackson    186
847   Movie       R           Tom Hooper    246
1524  Movie   TV-14          Umesh Mehra    228

In [571]:
#create a dictionary rating:director as key: value pairs
director_Dict = df_directors_count.set_index('rating').to_dict()['director']
director_Dict

{'66 min': 'Louis C.K.',
 '74 min': 'Louis C.K.',
 '84 min': 'Louis C.K.',
 'G': 'Robert Vince',
 'NC-17': 'Abdellatif Kechiche',
 'NR': 'Lars von Trier',
 'PG': 'Raja Gosnell',
 'PG-13': 'Peter Jackson',
 'R': 'Quentin Tarantino',
 'TV-14': 'Rob Seidenglanz',
 'TV-G': 'Kenny Ortega',
 'TV-MA': 'Alan Poul',
 'TV-PG': 'Noam Murro',
 'TV-Y': 'Guy Vasilovich',
 'TV-Y7': 'Thomas Astruc',
 'TV-Y7-FV': 'Mario Cambi',
 'UR': 'Sylvie Verheyde'}

In [572]:
#replace NaN values in director column based on dictionary values in rating column
df_exploded['director'] = df_exploded['director'].fillna(df_exploded['rating'].map(director_Dict))

# Here we can observe NaN values in director column are reduced from 50643 entries to 60 entries

In [573]:
df_exploded.isna().sum()

show_id             0
type                0
title               0
director           60
cast             2149
country         12003
date_added        158
release_year        0
rating             67
duration            3
listed_in           0
description         0
Day               158
month             158
dtype: int64

# Handling Missing Values for Cast Column

In [574]:
#create dataframe by grouping the objects type  and rating and get the value_counts for cast for each combination
df_cast=df_exploded.groupby(['type','rating'],as_index=False)['cast'].value_counts()

#for each group get the get the cast name whose made maximum movies in that particular group
df_cast_count=df_cast.loc[df_cast.reset_index().groupby(['type','rating'])['count'].idxmax()]

#create a dictionary rating:cast as key: value pairs
cast_Dict = df_cast_count.set_index('rating').to_dict()['cast']

#replace NaN values in cast column based on dictionary values in rating column
df_exploded['cast'] = df_exploded['cast'].fillna(df_exploded['rating'].map(cast_Dict))

df_exploded.isna().sum()

show_id             0
type                0
title               0
director           60
cast                0
country         12003
date_added        158
release_year        0
rating             67
duration            3
listed_in           0
description         0
Day               158
month             158
dtype: int64

# Here we can observe NaN values in cast column are reduced from  2149 entries to 0 entries

# Handling Missing Values for Country Column

In [577]:
#create dataframe by grouping the objects cast and get the value_counts for country for each combination: 
#Here we are identifying which actor belong to which country
df_country=df_exploded.groupby(['director'],as_index=False)['country'].value_counts()
#for each cast get the get the country name
df_country_count=df_country.loc[df_country.reset_index().groupby(['director'])['count'].idxmax()]

#create a dictionary cast:country as key: value pairs
country_Dict =df_country_count.set_index('director').to_dict()['country']

#replace NaN values in country column based on dictionary values in cast column
df_exploded['country'] = df_exploded['country'].fillna(df_exploded['director'].map(country_Dict))
df_exploded.isna().sum()

show_id            0
type               0
title              0
director          60
cast               0
country         4278
date_added       158
release_year       0
rating            67
duration           3
listed_in          0
description        0
Day              158
month            158
dtype: int64

# Here we can observe NaN values in country column are reduced from 12003 entries to 4278 entries

In [578]:
df_exploded['country'].nunique()

122

In [579]:
df_exploded['country'].unique()

array(['United States', 'South Africa', 'France', 'India', nan, 'Ghana',
       'Burkina Faso', 'United Kingdom', 'Germany', 'Ethiopia',
       'Czech Republic', 'Brazil', 'Mexico', 'Turkey', 'Australia',
       'Finland', 'China', 'Canada', 'Japan', 'Nigeria', 'Spain',
       'Belgium', 'South Korea', 'Singapore', 'Philippines', 'Italy',
       'Romania', 'Argentina', 'Venezuela', 'Hong Kong', 'Russia',
       'Ireland', 'Nepal', 'New Zealand', 'Greece', 'Jordan', 'Colombia',
       'Switzerland', 'Israel', 'Taiwan', 'Bulgaria', 'Algeria', 'Poland',
       'Egypt', 'Saudi Arabia', 'Thailand', 'Indonesia', 'Denmark',
       'Kuwait', 'Netherlands', 'Malaysia', 'Vietnam', 'Hungary',
       'Sweden', 'Lebanon', 'Syria', 'Iceland', 'United Arab Emirates',
       'Norway', 'Qatar', 'Mauritius', 'Austria', 'Cameroon', 'Palestine',
       'Uruguay', 'Kenya', 'Chile', 'Luxembourg', 'Cambodia',
       'Bangladesh', 'Portugal', 'Cayman Islands', 'Senegal', 'Serbia',
       'Malta', 'Namibia', '

# Handling Missing Values for Day  Column

In [580]:
##create dataframe by grouping the objects tyep and director and get the value_counts for day for each combination:
#Here we are identifying on which most occuring day(max value count) , the director of particular movie type released their movie
df_day=df_exploded.groupby(['type','director'],as_index=False)['Day'].value_counts()

#for each director get the  maximum count of day of release
df_day_count=df_day.loc[df_day.reset_index().groupby(['type','director'])['count'].idxmax()]

#create a dictionary director:day as key: value pairs
day_Dict =df_day_count.set_index('director').to_dict()['Day']

#replace NaN values in Day column based on dictionary values in director column
df_exploded['Day'] = df_exploded['Day'].fillna(df_exploded['director'].map(day_Dict))
df_exploded.isna().sum()

show_id            0
type               0
title              0
director          60
cast               0
country         4278
date_added       158
release_year       0
rating            67
duration           3
listed_in          0
description        0
Day                0
month            158
dtype: int64

In [581]:
df_exploded['Day'].unique()

array(['Saturday', 'Friday', 'Thursday', 'Wednesday', 'Tuesday', 'Monday',
       'Sunday'], dtype=object)

# Here we can observe NaN values in Day column are reduced from 158 entries to 0 entries

# Handling Missing Values for Month Column

In [582]:
##create dataframe by grouping the objects tyep and director and get the value_counts for day for each combination:
#Here we are identifying on which most occuring day(max value count) , the director of particular movie type released their movie
df_month=df_exploded.groupby(['type','director'],as_index=False)['month'].value_counts()

#for each director get the  maximum count of day of release
df_month_count=df_month.loc[df_month.reset_index().groupby(['type','director'])['count'].idxmax()]

#create a dictionary director:day as key: value pairs
month_Dict =df_month_count.set_index('director').to_dict()['month']

#replace NaN values in Day column based on dictionary values in director column
df_exploded['month'] = df_exploded['month'].fillna(df_exploded['director'].map(month_Dict)) 
df_exploded.isna().sum()

show_id            0
type               0
title              0
director          60
cast               0
country         4278
date_added       158
release_year       0
rating            67
duration           3
listed_in          0
description        0
Day                0
month              0
dtype: int64

In [584]:
df_exploded['month'].unique()

array(['September', 'August', 'July', 'June', 'May', 'April', 'March',
       'February', 'January', 'December', 'November', 'October'],
      dtype=object)

# Handling Missing Values for rating Column

In [585]:
#create dataframe by grouping the objects cast and get the value_counts for country for each combination: 
#Here we are identifying which actor belong to which country
df_country=df_exploded.groupby(['director','cast'],as_index=False)['rating'].value_counts()
#for each cast get the get the country name
df_country_count=df_country.loc[df_country.reset_index().groupby(['director','cast'])['count'].idxmax()]

#create a dictionary cast:country as key: value pairs
country_Dict =df_country_count.set_index('cast').to_dict()['rating']

#replace NaN values in country column based on dictionary values in cast column
df_exploded['rating'] = df_exploded['rating'].fillna(df_exploded['cast'].map(country_Dict))
df_exploded.isna().sum()

show_id            0
type               0
title              0
director          60
cast               0
country         4278
date_added       158
release_year       0
rating            15
duration           3
listed_in          0
description        0
Day                0
month              0
dtype: int64

In [586]:
print(round((df_exploded.isna().sum()/df_exploded.shape[0]),4)*100)

show_id         0.00
type            0.00
title           0.00
director        0.03
cast            0.00
country         2.12
date_added      0.08
release_year    0.00
rating          0.01
duration        0.00
listed_in       0.00
description     0.00
Day             0.00
month           0.00
dtype: float64


# Here we can observe that pecentage of null values  25% has reduced which are less than 2%. At this stage we can ignore, or drop the rows with NaN values

In [587]:
df_exploded.head(10)

show_id     type                 title         director            cast  \
0        1    Movie  Dick Johnson Is Dead  Kirsten Johnson  Luci Christian   
1        2  TV Show         Blood & Water        Alan Poul      Ama Qamata   
2        2  TV Show         Blood & Water        Alan Poul      Ama Qamata   
3        2  TV Show         Blood & Water        Alan Poul      Ama Qamata   
4        2  TV Show         Blood & Water        Alan Poul     Khosi Ngema   
5        2  TV Show         Blood & Water        Alan Poul     Khosi Ngema   
6        2  TV Show         Blood & Water        Alan Poul     Khosi Ngema   
7        2  TV Show         Blood & Water        Alan Poul   Gail Mabalane   
8        2  TV Show         Blood & Water        Alan Poul   Gail Mabalane   
9        2  TV Show         Blood & Water        Alan Poul   Gail Mabalane   

         country date_added  release_year rating   duration  \
0  United States 2021-09-25          2020  PG-13     90 min   
1   South Africa 2021-09-24          2021  TV-MA  2 Seasons   
2   South Africa 2021-09-24          2021  TV-MA  2 Seasons   
3   South Africa 2021-09-24          2021  TV-MA  2 Seasons   
4   South Africa 2021-09-24          2021  TV-MA  2 Seasons   
5   South Africa 2021-09-24          2021  TV-MA  2 Seasons   
6   South Africa 2021-09-24          2021  TV-MA  2 Seasons   
7   South Africa 2021-09-24          2021  TV-MA  2 Seasons   
8   South Africa 2021-09-24          2021  TV-MA  2 Seasons   
9   South Africa 2021-09-24          2021  TV-MA  2 Seasons   

                listed_in                                        description  \
0           Documentaries  As her father nears the end of his life, filmm...   
1  International TV Shows  After crossing paths at a party, a Cape Town t...   
2               TV Dramas  After crossing paths at a party, a Cape Town t...   
3            TV Mysteries  After crossing paths at a party, a Cape Town t...   
4  International TV Shows  After crossing paths at a party, a Cape Town t...   
5               TV Dramas  After crossing paths at a party, a Cape Town t...   
6            TV Mysteries  After crossing paths at a party, a Cape Town t...   
7  International TV Shows  After crossing paths at a party, a Cape Town t...   
8               TV Dramas  After crossing paths at a party, a Cape Town t...   
9            TV Mysteries  After crossing paths at a party, a Cape Town t...   

        Day      month  
0  Saturday  September  
1    Friday  September  
2    Friday  September  
3    Friday  September  
4    Friday  September  
5    Friday  September  
6    Friday  September  
7    Friday  September  
8    Friday  September  
9    Friday  September

# DATA ANALYSIS

# 1.How has the number of movies released per year changed over the last 20-30 years?

In [591]:
movies_released=df_exploded.groupby('release_year')['type'].value_counts().reset_index(name='counts')

In [592]:
movies_released

release_year     type  counts
0            1925  TV Show       1
1            1942    Movie       6
2            1943    Movie       5
3            1944    Movie      24
4            1945    Movie      12
..            ...      ...     ...
114          2019  TV Show   10005
115          2020  TV Show   10089
116          2020    Movie    9608
117          2021  TV Show    6467
118          2021    Movie    5427

[119 rows x 3 columns]

In [594]:
recent_movies=movies_released.loc[movies_released['release_year']>=1990]
recent_movies

release_year     type  counts
55           1990    Movie     422
56           1990  TV Show      98
57           1991    Movie     532
58           1991  TV Show       1
59           1992    Movie     504
..            ...      ...     ...
114          2019  TV Show   10005
115          2020  TV Show   10089
116          2020    Movie    9608
117          2021  TV Show    6467
118          2021    Movie    5427

[64 rows x 3 columns]

In [595]:
fig = px.line(recent_movies, x="release_year", y="counts", color='type',
              title='', 
             markers=True)
fig.update_layout(xaxis_title="Release year",
                  yaxis_title="Number of content",
                  legend_title='Type of Content')


# 2.Comparison of tv shows vs. movies.

In [596]:
df_tvshows_movies=df_exploded.groupby(['type'],as_index=False)['type'].value_counts()

In [597]:
df_tvshows_movies

type   count
0    Movie  145917
1  TV Show   56148

In [598]:
px.bar(data_frame=df_moviesreleased_year, x="type", y="count", color="type", barmode="group",title="Total Contents available in Netflix")

# 3. What is the best time(day) to launch a TV show?

In [599]:
df_bestday=df_exploded.groupby(['Day'],as_index=False)['type'].count()
#df_day=df_bestday.groupby(['Day']).sum()

In [600]:
df_bestday

Day   type
0     Friday  58185
1     Monday  18987
2   Saturday  19278
3     Sunday  19246
4   Thursday  33432
5    Tuesday  24003
6  Wednesday  28934

In [601]:
fig = px.bar(df_bestday, x="Day", y="type", color='Day',
              title='')
fig.update_layout(xaxis_title="Release Day",
                  yaxis_title="Number of movies released",
                  legend_title='days')

# 4. Analysis of actors/directors of different types of shows/movies.

#    4. a) Most casted Actor in Netflix content

In [602]:
df_mostcastactor=df_exploded.groupby(['cast'],as_index=False)['type'].value_counts().sort_values(by='count',ascending=False)

In [603]:
df_mostcast=df_mostcastactor.head(10)

In [604]:
df_mostcast

cast     type  count
17337        Joanna Kulig    Movie    511
17338        Joanna Kulig  TV Show    417
804             Ai Kayano    Movie    349
8817   David Attenborough    Movie    306
8818   David Attenborough  TV Show    270
805             Ai Kayano  TV Show    225
1492        Alfred Molina    Movie    171
22126         Liam Neeson    Movie    161
17711      John Krasinski    Movie    138
32976         Salma Hayek    Movie    130

In [605]:
fig = px.line(df_mostcast, x="cast", y="count", color='type',
              title='', 
             markers=True)
fig.update_layout(xaxis_title="cast",
                  yaxis_title="Number of movies released",
                  legend_title='Type of Content')

#   4. b)Top director contributed to the netflix content

In [606]:
df_mostcastdirec=df_exploded.groupby(['director'],as_index=False)['director'].value_counts().sort_values(by='count',ascending=False)

In [607]:
df_topdirec=df_mostcastdirec.head(10)

In [608]:
df_topdirec

director  count
140             Alan Poul  25945
3899      Rob Seidenglanz  13532
3380           Noam Murro   4528
4590        Thomas Astruc   3992
1640       Guy Vasilovich   1870
2461         Kenny Ortega   1136
2605       Lars von Trier    491
2945      Martin Scorsese    419
4941      Youssef Chahine    409
755   Cathy Garcia-Molina    356

In [609]:
fig = px.bar(df_topdirec, x="director", y="count", color='director',
              title='')
fig.update_layout(xaxis_title="director",
                  yaxis_title="Number of movies released",
                  legend_title='top directors')

# 4.C)Top directors in different genre (considered top 20 genres)

In [610]:
df_directorby_genre=df_exploded.groupby(['listed_in','director'],as_index=False)['director'].value_counts()
df_directorby_genre

listed_in                  director  count
0      Action & Adventure                 A. Salaam     10
1      Action & Adventure           A.R. Murugadoss     12
2      Action & Adventure                 Aaron Nee     13
3      Action & Adventure  Abbas Alibhai Burmawalla     24
4      Action & Adventure               Abhinay Deo     19
...                   ...                       ...    ...
12250           Thrillers               Zack Whedon     24
12251           Thrillers              Zak Hilditch     13
12252           Thrillers                 Zeek Earl     12
12253           Thrillers               Zoya Akhtar     10
12254           Thrillers               Àlex Pastor     12

[12255 rows x 3 columns]

In [611]:
df_director_genre_count=df_directorby_genre.loc[df_directorby_genre.reset_index().groupby(['listed_in'])['count'].idxmax()]
top_directors=df_director_genre_count.head(20).sort_values(by='count',ascending=False)
top_directors

listed_in           director  count
9429    International TV Shows          Alan Poul   6629
2875            Crime TV Shows          Alan Poul   3146
9609                  Kids' TV      Thomas Astruc   2292
802           British TV Shows          Alan Poul    878
786               Anime Series          Alan Poul    859
9626           Korean TV Shows    Rob Seidenglanz    456
3897                Docuseries          Alan Poul    361
5147                    Dramas    Martin Scorsese    163
389         Action & Adventure    Martin Campbell    154
2507                  Comedies       Raja Gosnell    154
1231  Children & Family Movies       Raja Gosnell    154
9387      International Movies    Youssef Chahine    132
6831        Independent Movies     Lars von Trier    112
1379         Classic & Cult TV    Rob Seidenglanz     93
1473            Classic Movies    Youssef Chahine     89
772             Anime Features  Toshiya Shinohara     68
6021      Faith & Spirituality        David Batty     63
6198             Horror Movies          James Wan     61
2975               Cult Movies       Edgar Wright     56
3046             Documentaries          Alan Poul     49

In [612]:
fig = px.bar(top_directors, x="director", y="count", color='listed_in',
              title='')
fig.update_layout(xaxis_title="directors",
                  yaxis_title="Number of movies released",
                  legend_title='Top 20 genres')

# 4.d Top Actors in different genre

In [613]:
df_actor_genre=df_exploded.groupby(['listed_in','cast'],as_index=False)['cast'].value_counts()
df_actor_genre
df_actor_genre_count=df_actor_genre.loc[df_actor_genre.reset_index().groupby(['listed_in'])['count'].idxmax()]
top_actors=df_actor_genre_count.head(20).sort_values(by='count',ascending=False)
top_actors

listed_in                cast  count
28212             Documentaries        Joanna Kulig    261
57584      International Movies        Joanna Kulig    130
29049                Docuseries        Joanna Kulig    128
68928    International TV Shows        Joanna Kulig     80
24965            Crime TV Shows        Joanna Kulig     70
10138  Children & Family Movies      John Krasinski     66
29900                    Dramas       Alfred Molina     65
75531                  Kids' TV        Kate Harbour     40
7402           British TV Shows  David Attenborough     40
22520                  Comedies         Tara Strong     31
3114         Action & Adventure      Luci Christian     28
45519             Horror Movies        Lorenza Izzo     25
5801             Anime Features         Andy McAvin     24
6970               Anime Series    Takahiro Sakurai     24
48831        Independent Movies        James Franco     19
27397               Cult Movies         Keith David     14
12504            Classic Movies    Burgess Meredith     11
43535      Faith & Spirituality     Abdelilah Wahbi      8
76705           Korean TV Shows           Bae Doona      6
12212         Classic & Cult TV        Don S. Davis      3

In [614]:
fig = px.bar(top_actors, x="cast", y="count", color='listed_in',
              title='')
fig.update_layout(xaxis_title="cast",
                  yaxis_title="Number of movies released",
                  legend_title='Top 20 genres')

# 4.e Successful Actor-Director-genre combo in Netlfix

In [615]:
df_actor_dir_genre=df_exploded.groupby(['listed_in','director','cast'],as_index=False)['director','cast'].value_counts()
df_actor_dir_genre
df_actor_dir_genre_count=df_actor_dir_genre.loc[df_actor_dir_genre.reset_index().groupby(['listed_in'])['count'].idxmax()]
top_actor_dir_genre=df_actor_dir_genre_count.head(20).sort_values(by='count',ascending=False)
top_actor_dir_genre

C:\Users\Anne\AppData\Local\Temp\ipykernel_11496\2826017237.py:1: FutureWarning:

Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.



listed_in           director                  cast  \
41969                 Docuseries          Alan Poul          Joanna Kulig   
97718     International TV Shows          Alan Poul          Joanna Kulig   
36490             Crime TV Shows          Alan Poul          Joanna Kulig   
106182                  Kids' TV     Guy Vasilovich          Kate Harbour   
11305           British TV Shows         Noam Murro    David Attenborough   
15953   Children & Family Movies      Rajiv Chilaka         Julie Tejwani   
9304                Anime Series          Alan Poul      Takahiro Sakurai   
40071              Documentaries          Alan Poul          Joanna Kulig   
86206       International Movies     Matthew Salleh          Joanna Kulig   
71359         Independent Movies     Lars von Trier  Charlotte Gainsbourg   
53038                     Dramas     Lars von Trier  Charlotte Gainsbourg   
7875          Action & Adventure         Wilson Yip            Donnie Yen   
30968                   Comedies       Raja Gosnell          Alan Cumming   
63872       Faith & Spirituality        David Batty       Abdelilah Wahbi   
8801              Anime Features  Toshiya Shinohara       Houko Kuwashima   
65545              Horror Movies          James Wan        Patrick Wilson   
39417                Cult Movies       Edgar Wright           Alison Pill   
18176             Classic Movies         David Lean         Alec Guinness   
108669           Korean TV Shows          Alan Poul          Seo Hyun-jin   
17831          Classic & Cult TV          Alan Poul        Amanda Tapping   

        count  
41969      87  
97718      48  
36490      43  
106182     35  
11305      24  
15953      19  
9304       14  
40071      13  
86206      12  
71359      11  
53038      11  
7875       10  
30968      10  
63872       8  
8801        7  
65545       6  
39417       4  
18176       4  
108669      4  
17831       2

# 5. Does Netflix has more focus on TV Shows than movies in recent years

In [616]:
df_netflix_focus=df_exploded.groupby(['release_year','type'],as_index=False)['type'].value_counts()

In [617]:
netflix_focus=df_netflix_focus.sort_values(by='release_year',ascending=False).head(40)
netflix_focus

release_year     type  count
118          2021  TV Show   6467
117          2021    Movie   5427
116          2020  TV Show  10089
115          2020    Movie   9608
114          2019  TV Show  10005
113          2019    Movie  11926
112          2018  TV Show   7381
111          2018    Movie  17060
110          2017  TV Show   5447
109          2017    Movie  15069
108          2016  TV Show   4390
107          2016    Movie  14075
106          2015  TV Show   3516
105          2015    Movie  10612
104          2014  TV Show   1454
103          2014    Movie   7644
102          2013  TV Show   1223
101          2013    Movie   6522
100          2012  TV Show   1077
99           2012    Movie   5277
98           2011  TV Show    599
97           2011    Movie   3974
96           2010  TV Show    709
95           2010    Movie   4400
94           2009  TV Show    581
93           2009    Movie   4201
92           2008  TV Show    622
91           2008    Movie   3169
89           2007    Movie   2231
90           2007  TV Show    259
88           2006  TV Show    318
87           2006    Movie   2205
86           2005  TV Show    205
85           2005    Movie   2032
84           2004  TV Show    175
83           2004    Movie   1906
82           2003  TV Show    242
81           2003    Movie   1692
80           2002  TV Show    164
79           2002    Movie   1613

In [618]:
fig = px.line(netflix_focus, x="release_year", y="count", color='type',
              title='', 
             markers=True)
fig.update_layout(xaxis_title="year",
                  yaxis_title="Number of movies released",
                  legend_title='Type of Content')

# 6.Understanding what content is available in different countries

In [695]:
df_content_in_countries=df_exploded.groupby(['country','rating'],as_index=False)['rating'].value_counts()
df_content_in_countries

country rating  count
0     Afghanistan  TV-MA      2
1         Albania  TV-MA      8
2         Algeria  TV-14     84
3         Algeria  TV-MA     53
4          Angola  TV-MA     32
..            ...    ...    ...
509  West Germany  TV-14     15
510  West Germany  TV-MA     46
511      Zimbabwe     NR      2
512      Zimbabwe   TV-G     36
513      Zimbabwe  TV-MA      4

[514 rows x 3 columns]

In [696]:
df_content_in_countries_count=df_content_in_countries.loc[df_content_in_countries.reset_index().groupby(['country'])['count'].idxmax()]
top_content_in_countries=df_content_in_countries_count.head(20).sort_values(by='count',ascending=False)
top_content_in_countries
#df_content_in_countries_count

country rating  count
11     Argentina  TV-MA   1448
55        Brazil  TV-MA   1187
43       Belgium  TV-MA   1040
21     Australia  TV-MA    724
33    Bangladesh  TV-MA    330
59      Bulgaria      R    140
27       Austria  TV-MA     94
2        Algeria  TV-14     84
29    Azerbaijan  TV-14     33
4         Angola  TV-MA     32
66      Cambodia  TV-14     26
62  Burkina Faso  TV-MA     24
68      Cameroon  TV-MA     20
47       Bermuda  TV-PG     12
1        Albania  TV-MA      8
35       Belarus   TV-Y      6
30       Bahamas  PG-13      6
48      Botswana  TV-PG      2
14       Armenia  TV-MA      2
0    Afghanistan  TV-MA      2

In [706]:
fig = px.bar(top_content_in_countries, x="country", y="rating", color='rating',
              title='')
fig.update_layout(xaxis_title="country",
                  yaxis_title="type",
                  legend_title='Type of Content')

# 8.Top Genre wise movie and TV shows released by year

In [622]:
df_genre_type_release_year=df_exploded.groupby(['release_year','type','listed_in'],as_index=False)['listed_in'].value_counts()
df_genre_type_release_year

release_year     type            listed_in  count
0             1925  TV Show             TV Shows      1
1             1942    Movie       Classic Movies      3
2             1942    Movie        Documentaries      3
3             1943    Movie       Classic Movies      1
4             1943    Movie        Documentaries      4
...            ...      ...                  ...    ...
1187          2021  TV Show         TV Mysteries    205
1188          2021  TV Show  TV Sci-Fi & Fantasy    172
1189          2021  TV Show             TV Shows    215
1190          2021  TV Show         TV Thrillers    101
1191          2021  TV Show        Teen TV Shows     76

[1192 rows x 4 columns]

In [623]:
df_genre_count=df_genre_type_release_year.loc[df_genre_type_release_year.reset_index().groupby(['listed_in'])['count'].idxmax()]
top_genre_in_year=df_genre_count.head(20).sort_values(by='release_year',ascending=False)
top_genre_in_year
#df_content_in_countries_count

release_year     type                 listed_in  count
1171          2021  TV Show              Anime Series    336
1174          2021  TV Show                Docuseries    173
1116          2020    Movie  Children & Family Movies   1019
1137          2020  TV Show    International TV Shows   2431
1093          2019  TV Show          British TV Shows    359
1094          2019  TV Show         Classic & Cult TV     43
1098          2019  TV Show                  Kids' TV    709
1095          2019  TV Show            Crime TV Shows   1012
1043          2018    Movie        Independent Movies   1535
1037          2018    Movie                  Comedies   2155
1034          2018    Movie            Anime Features    177
1040          2018    Movie                    Dramas   3673
1041          2018    Movie      Faith & Spirituality    134
1044          2018    Movie      International Movies   3587
992           2017    Movie        Action & Adventure   1226
1018          2017  TV Show           Korean TV Shows    175
960           2016    Movie             Horror Movies    652
957           2016    Movie             Documentaries    413
807           2012    Movie               Cult Movies    140
325           1992    Movie            Classic Movies    114

In [624]:
fig = px.bar(top_genre_in_year, x="listed_in",
                   y="count",
                   title="Genre wise and Category (Movie / TV Shows) wise content added in Netflix",
                   color='type')
fig.show()

# 9. Top 20 countries contributing to netflix by type

In [625]:
df_top_countries=df_exploded.groupby(['type','country'],as_index=False)['country'].value_counts().sort_values(by='count',ascending=False)
topcountries1=df_top_countries.head(20)
topcountries1

type         country  count
110    Movie   United States  46391
41     Movie           India  22283
179  TV Show   United States  18280
109    Movie  United Kingdom   8609
32     Movie          France   6641
18     Movie          Canada   5755
146  TV Show           Japan   5154
178  TV Show  United Kingdom   4388
168  TV Show     South Korea   3754
49     Movie           Japan   3710
97     Movie           Spain   3533
34     Movie         Germany   3443
73     Movie         Nigeria   2486
21     Movie           China   2377
38     Movie       Hong Kong   2205
124  TV Show          Canada   2177
29     Movie           Egypt   2173
42     Movie       Indonesia   2089
154  TV Show          Mexico   2018
63     Movie          Mexico   1967

In [626]:
fig = px.pie(topcountries, values='count', names='country',title="Top 20 Contries Contributing to Netflix")
fig.update_layout(xaxis_title="country",
                  yaxis_title="count",
                  legend_title='countries',
                  height=600)
fig.show()

# 10.Movies Vs TV shows released in last 5 years- Rating wise

In [637]:
movies_vs_tv_rating=df_exploded.groupby(['release_year','rating','type'],as_index=False)['type'].value_counts().sort_values(by='release_year',ascending=False)
movies_vs_tv_ratingwise=movies_vs_tv_rating.loc[movies_vs_tv_rating['release_year']>=2015]
movies_vs_tv_ratingwise

release_year rating     type  count
565          2021  TV-Y7  TV Show    294
558          2021  TV-MA    Movie   2340
551          2021     PG    Movie    321
552          2021  PG-13    Movie    243
553          2021      R    Movie    360
..            ...    ...      ...    ...
458          2015   TV-Y    Movie    313
459          2015   TV-Y  TV Show    170
460          2015  TV-Y7    Movie     34
461          2015  TV-Y7  TV Show    141
455          2015  TV-MA  TV Show   1414

[123 rows x 4 columns]

In [668]:
fig = px.bar(movies_vs_tv_ratingwise, x="type", y="rating", color='rating',
              title='')
fig.update_layout(xaxis_title="type",
                  yaxis_title="genre",
                  legend_title='Type of Content')

# 11.Movies Vs TV shows released in last 5 years- Genre wise

In [657]:
movies_vs_tv_genre=df_exploded.groupby(['release_year','listed_in','type'],as_index=False)['type'].value_counts().sort_values(by='release_year',ascending=False)
movies_vs_tv_genrewise=movies_vs_tv_genre.loc[movies_vs_tv_genre['release_year']>=2015]
movies_vs_tv_genrewise

release_year               listed_in     type  count
1191          2021               Thrillers    Movie    325
1162          2021              Docuseries  TV Show    173
1170          2021         Korean TV Shows  TV Show     30
1169          2021                Kids' TV  TV Show    378
1168          2021  International TV Shows  TV Show   1466
...            ...                     ...      ...    ...
936           2015        Sci-Fi & Fantasy    Movie    257
937           2015     Science & Nature TV  TV Show      9
935           2015       Romantic TV Shows  TV Show    260
939           2015           Sports Movies    Movie    124
940           2015         Stand-Up Comedy    Movie     18

[280 rows x 4 columns]

In [667]:
fig = px.bar(movies_vs_tv_genrewise, x="type", y="listed_in", color='listed_in',
              title='')
fig.update_layout(xaxis_title="type",
                  yaxis_title="genre",
                  legend_title='Type of Content')

# 12. Most content released in month wise

In [627]:
df_top_content=df_exploded.groupby(['month','listed_in'],as_index=False)['listed_in'].value_counts().sort_values(by='month',ascending=False)
df_top_content


month                  listed_in  count
498  September                  Thrillers    768
466  September                Cult Movies    177
475  September                   Kids' TV    416
474  September     International TV Shows    947
473  September       International Movies   2619
..         ...                        ...    ...
27       April  Spanish-Language TV Shows    208
26       April        Science & Nature TV     12
25       April           Sci-Fi & Fantasy    316
24       April          Romantic TV Shows    251
0        April         Action & Adventure    989

[499 rows x 3 columns]

In [630]:
df_content_count=df_top_content.loc[df_top_content.reset_index().groupby(['month'])['count'].idxmax()]
top_content_in_month=df_content_count.sort_values(by='month',ascending=False)
top_content_in_month

month                 listed_in  count
458  September            Anime Features    236
420    October  Children & Family Movies   1294
375   November        Action & Adventure   1016
341        May            Crime TV Shows    363
298      March            Classic Movies     86
257       June                  Comedies   1380
215       July                  Comedies   2074
173    January                  Comedies   2244
130   February            Classic Movies      9
92    December               Cult Movies     60
44      August          British TV Shows     85
4        April  Children & Family Movies    648

In [631]:
fig = px.bar(top_content_in_month, x="month", y="count", color='listed_in',
              title='')
fig.update_layout(xaxis_title="montht",
                  yaxis_title="Number of movies released",
                  legend_title='genres')

# 6. Insights based on Non-Graphical and Visual Analysis 

# 6.1 Comments on the range of attributes

In [669]:
df_exploded.head(5)

show_id     type                 title         director            cast  \
0        1    Movie  Dick Johnson Is Dead  Kirsten Johnson  Luci Christian   
1        2  TV Show         Blood & Water        Alan Poul      Ama Qamata   
2        2  TV Show         Blood & Water        Alan Poul      Ama Qamata   
3        2  TV Show         Blood & Water        Alan Poul      Ama Qamata   
4        2  TV Show         Blood & Water        Alan Poul     Khosi Ngema   

         country date_added  release_year rating   duration  \
0  United States 2021-09-25          2020  PG-13     90 min   
1   South Africa 2021-09-24          2021  TV-MA  2 Seasons   
2   South Africa 2021-09-24          2021  TV-MA  2 Seasons   
3   South Africa 2021-09-24          2021  TV-MA  2 Seasons   
4   South Africa 2021-09-24          2021  TV-MA  2 Seasons   

                listed_in                                        description  \
0           Documentaries  As her father nears the end of his life, filmm...   
1  International TV Shows  After crossing paths at a party, a Cape Town t...   
2               TV Dramas  After crossing paths at a party, a Cape Town t...   
3            TV Mysteries  After crossing paths at a party, a Cape Town t...   
4  International TV Shows  After crossing paths at a party, a Cape Town t...   

        Day      month  
0  Saturday  September  
1    Friday  September  
2    Friday  September  
3    Friday  September  
4    Friday  September

In [670]:
df_exploded['type'].unique()

array(['Movie', 'TV Show'], dtype=object)

# For attribute "type": type has two values Movie and TV Show, which reresent the type of show

In [671]:
df_exploded['title'].nunique()

8807

# For attribute "title": Dataset contains 8807 unique show titles together from Movies and TV shows

In [672]:
df_exploded['director'].nunique()

4993

# For attribute "director": Dataset contains 4993 unique directors together from Movies and TV shows

In [674]:
df_exploded['cast'].nunique()

36440

# For attribute "cast": Dataset contains 36440 unique actors together from Movies and TV shows¶

In [675]:
df_exploded['country'].nunique()

122

# For attribute country": Dataset contains 122 unique countries. i.e data is collected for 122 countries

In [676]:
df_exploded['release_year'].unique()

array([2020, 2021, 1993, 2018, 1996, 1998, 1997, 2010, 2013, 2017, 1975,
       1978, 1983, 1987, 2012, 2001, 2014, 2002, 2003, 2004, 2011, 2008,
       2009, 2007, 2005, 2006, 1994, 2015, 2019, 2016, 1982, 1989, 1990,
       1991, 1999, 1986, 1992, 1984, 1980, 1961, 2000, 1995, 1985, 1976,
       1959, 1988, 1981, 1972, 1964, 1945, 1954, 1979, 1958, 1956, 1963,
       1970, 1973, 1925, 1974, 1960, 1966, 1971, 1962, 1969, 1977, 1967,
       1968, 1965, 1946, 1942, 1955, 1944, 1947, 1943], dtype=int64)

# For attribute "release_year": Dataset is taken from the year 1943 to 2021

In [677]:
df_exploded['rating'].nunique()

17

In [680]:
df_exploded['duration'].nunique()

220

In [681]:
df_exploded['listed_in'].nunique()

42

# 6.2 Comments on the distribution of the variables and relationship between them

In [683]:
df_exploded.corr()

C:\Users\Anne\AppData\Local\Temp\ipykernel_11496\1617123906.py:1: FutureWarning:

The default value of numeric_only in DataFrame.corr is deprecated. In a future version, it will default to False. Select only valid columns or specify the value of numeric_only to silence this warning.



show_id  release_year
show_id       1.00000      -0.23285
release_year -0.23285       1.00000

In [693]:
pd.crosstab(df_exploded.type,df_exploded.duration)


duration  1 Season  10 Seasons  10 min  100 min  101 min  102 min  103 min  \
type                                                                         
Movie            0           0       4     2699     3048     3017     2985   
TV Show      35035         220       0        0        0        0        0   

duration  104 min  105 min  106 min  ...  90 min  91 min  92 min  93 min  \
type                                 ...                                   
Movie        2822     3209     4040  ...    3305    2915    2863    3480   
TV Show         0        0        0  ...       0       0       0       0   

duration  94 min  95 min  96 min  97 min  98 min  99 min  
type                                                      
Movie       4343    3560    3511    3624    2984    2956  
TV Show        0       0       0       0       0       0  

[2 rows x 220 columns]

In [689]:
pd.crosstab(df_exploded.type,df_exploded.rating)

rating   66 min  74 min  84 min     G  NC-17    NR     PG  PG-13      R  \
type                                                                      
Movie         1       1       1  1530    149  1418  10919  16246  25806   
TV Show       0       0       0     0      0   155      0      2     54   

rating   TV-14  TV-G  TV-MA  TV-PG  TV-Y  TV-Y7  TV-Y7-FV  UR  
type                                                           
Movie    29266  1738  44009  10314  1878   2486        62  86  
TV Show  14705  1041  29906   4616  1789   3848        24   0

In [691]:
pd.crosstab(df_exploded.type,df_exploded.listed_in)

listed_in  Action & Adventure  Anime Features  Anime Series  British TV Shows  \
type                                                                            
Movie                   12216            1045             0                 0   
TV Show                     0               0          2313              1808   

listed_in  Children & Family Movies  Classic & Cult TV  Classic Movies  \
type                                                                     
Movie                          9771                  0            1443   
TV Show                           0                272               0   

listed_in  Comedies  Crime TV Shows  Cult Movies  ...  TV Action & Adventure  \
type                                              ...                          
Movie         20829               0         1077  ...                      0   
TV Show           0            4733            0  ...                   2288   

listed_in  TV Comedies  TV Dramas  TV Horror  TV Mysteries  \
type                                                         
Movie                0          0          0             0   
TV Show           4963       8942        941          1281   

listed_in  TV Sci-Fi & Fantasy  TV Shows  TV Thrillers  Teen TV Shows  \
type                                                                    
Movie                        0         0             0              0   
TV Show                   1045       337           768            742   

listed_in  Thrillers  
type                  
Movie           7107  
TV Show            0  

[2 rows x 42 columns]

# 7. Business Insights- patterns observed in the data along with inference

Observations:
1.	Over the past 23-30 years, maximum movies and tv shows are released during the years 2018 and 2019.
2.	More movies are released than Tv shows
3.	Maximum shows are released on Fridays
4.	Joanna Kuling, Ai Kayano and David Attenborough are the top 3 actors in the Netflix content
5.	Alan Poul, Rob Seidenglanz, Noam Murro are the top 3 directors in the netflic content
6.	Alan Poul is the top director from different genres
7.	Joanna Kuling is the top actor from different genres
8.	Up to year 2018. Netflix had more focus on Movies. 
9.	From year 2019. Netflix focused on releasing more tv shoes than movies
10.	TV-MA is the top content avaible in many countries. (Programming rated TV-MA in the United States by the TV Parental Guidelines signifies content for mature audiences.)
11.	United States and India are the top 1 and 2 countries respectively which are contributing to netflix

# 8. Recommendations - Actionable items for business

1.	Netflix can focus on releasing more movies
2.	The best day to release show is Friday
3.	For maximum profits, producers can opt to cast the actors Joanna Kuling, Ai Kayano and David Attenborough in their    shows
4.	For maximum profits, producers can opt to the directors Alan Poul, Rob Seidenglanz, Noam Murro  in their shows
5.	Producer can opt for director Alan Poul, as he is the top most director in multiple genres
6.	Producer can opt for actor Joanna Kuling, as he is the top most actor in multiple genres
7.	Alan Poul and Joanna Kuling is the best combination to produce a successful movie
8.	Over the recent years from 2019-2021, Netflix has more focus on tv shows
9.	Producer can focus on producing TV shows with TV-MA rating as this is highest contenet avaible in multiple countries.
10.	Producer can focus on producing shows in united states and india as they are highly contributing to Netflix